File: src/lib/tauri/jira.ts
import { invoke } from "@tauri-apps/api/core";
import { isMockMode } from "./core";
import type { BitbucketPr } from "./bitbucket";

// ── JIRA types ────────────────────────────────────────────────────────────────

export interface JiraSprint {
  id: number;
  name: string;
  state: string;
  startDate: string | null;
  endDate: string | null;
  completeDate: string | null;
  goal: string | null;
}

export interface JiraUser {
  accountId: string;
  displayName: string;
  emailAddress: string | null;
}

export interface DescriptionSection {
  heading: string | null;
  content: string;
}

export interface JiraIssue {
  id: string;
  key: string;
  url: string;
  summary: string;
  description: string | null;
  descriptionSections: DescriptionSection[];
  status: string;
  statusCategory: string;
  assignee: JiraUser | null;
  reporter: JiraUser | null;
  issueType: string;
  priority: string | null;
  storyPoints: number | null;
  labels: string[];
  epicKey: string | null;
  epicSummary: string | null;
  created: string;
  updated: string;
  resolutionDate: string | null;
  completedInSprint: boolean | null;
  /** Auto-detected from custom field display name — no configuration required. */
  acceptanceCriteria: string | null;
  stepsToReproduce: string | null;
  observedBehavior: string | null;
  expectedBehavior: string | null;
  /**
   * All non-empty custom fields keyed by human-readable display name.
   * Only populated by get_issue (full detail fetch). Empty for list/sprint fetches.
   */
  namedFields: Record<string, string>;
  /**
   * Mapping of semantic field name → discovered JIRA field ID.
   * e.g. { "acceptance_criteria": "customfield_10034" }
   * Empty when fields were not auto-discovered. Only populated by get_issue.
   */
  discoveredFieldIds: Record<string, string>;
}

// ── JIRA commands ─────────────────────────────────────────────────────────────

export async function getActiveSprint(): Promise<JiraSprint | null> {
  if (isMockMode()) {
    const { ACTIVE_SPRINT } = await import("../mockData/sprints");
    return ACTIVE_SPRINT;
  }
  return invoke<JiraSprint | null>("get_active_sprint");
}

export async function getAllActiveSprints(): Promise<JiraSprint[]> {
  if (isMockMode()) {
    const { ACTIVE_SPRINT } = await import("../mockData/sprints");
    return ACTIVE_SPRINT ? [ACTIVE_SPRINT] : [];
  }
  return invoke<JiraSprint[]>("get_all_active_sprints");
}

export async function getAllActiveSprintIssues(): Promise<
  Array<[JiraSprint, JiraIssue[]]>
> {
  if (isMockMode()) {
    const { ACTIVE_SPRINT, ACTIVE_SPRINT_2 } = await import("../mockData/sprints");
    const { SPRINT_ISSUES_BY_ID } = await import("../mockData/issues");
    if (!ACTIVE_SPRINT) return [];
    return [
      [ACTIVE_SPRINT, SPRINT_ISSUES_BY_ID[23] ?? []],
      [ACTIVE_SPRINT_2, SPRINT_ISSUES_BY_ID[24] ?? []],
    ];
  }
  return invoke<Array<[JiraSprint, JiraIssue[]]>>(
    "get_all_active_sprint_issues",
  );
}

export async function getActiveSprintIssues(): Promise<JiraIssue[]> {
  if (isMockMode()) {
    const { SPRINT_ISSUES_BY_ID } = await import("../mockData/issues");
    return SPRINT_ISSUES_BY_ID[23] ?? [];
  }
  return invoke<JiraIssue[]>("get_active_sprint_issues");
}

export async function getMySprintIssues(): Promise<JiraIssue[]> {
  if (isMockMode()) {
    const { MY_SPRINT_ISSUES } = await import("../mockData/issues");
    return MY_SPRINT_ISSUES;
  }
  return invoke<JiraIssue[]>("get_my_sprint_issues");
}

export async function getSprintIssues(sprintId: number): Promise<JiraIssue[]> {
  if (isMockMode()) {
    const { SPRINT_ISSUES_BY_ID } = await import("../mockData/issues");
    return SPRINT_ISSUES_BY_ID[sprintId] ?? [];
  }
  return invoke<JiraIssue[]>("get_sprint_issues", { sprintId });
}

export async function getSprintIssuesById(
  sprintId: number,
  completeDate?: string | null,
): Promise<JiraIssue[]> {
  if (isMockMode()) {
    const { SPRINT_ISSUES_BY_ID } = await import("../mockData/issues");
    return SPRINT_ISSUES_BY_ID[sprintId] ?? [];
  }
  return invoke<JiraIssue[]>("get_sprint_issues_by_id", { sprintId, completeDate: completeDate ?? null });
}

export async function getIssue(issueKey: string): Promise<JiraIssue> {
  if (isMockMode()) {
    const { ALL_ISSUES_BY_KEY } = await import("../mockData/issues");
    const issue = ALL_ISSUES_BY_KEY[issueKey];
    if (!issue) throw new Error(`Mock: issue ${issueKey} not found`);
    return issue;
  }
  return invoke<JiraIssue>("get_issue", { issueKey });
}

export async function getCompletedSprints(
  limit: number,
): Promise<JiraSprint[]> {
  if (isMockMode()) {
    const { COMPLETED_SPRINTS } = await import("../mockData/sprints");
    return COMPLETED_SPRINTS.slice(0, limit);
  }
  return invoke<JiraSprint[]>("get_completed_sprints", { limit });
}

// ── Sprint report disk cache ───────────────────────────────────────────────────

export interface SprintReportCache {
  issues: JiraIssue[];
  prs: BitbucketPr[];
  cachedAt: string;
  aiSummary?: string;
  aiSummaryGeneratedAt?: string;
}

export async function saveSprintReport(
  sprintId: number,
  data: SprintReportCache,
): Promise<void> {
  if (isMockMode()) return;
  return invoke<void>("save_sprint_report", {
    sprintId,
    dataJson: JSON.stringify(data),
  });
}

export async function loadSprintReport(
  sprintId: number,
): Promise<SprintReportCache | null> {
  if (isMockMode()) return null;
  const raw = await invoke<string | null>("load_sprint_report", { sprintId });
  if (!raw) return null;
  try {
    return JSON.parse(raw) as SprintReportCache;
  } catch {
    return null;
  }
}

export async function listCachedSprintIds(): Promise<number[]> {
  if (isMockMode()) return [];
  return invoke<number[]>("list_cached_sprint_ids");
}

export async function getSprintReportsDir(): Promise<string> {
  return invoke<string>("get_sprint_reports_dir");
}

export async function getFutureSprints(limit: number): Promise<JiraSprint[]> {
  if (isMockMode()) return [];
  return invoke<JiraSprint[]>("get_future_sprints", { limit });
}

export async function searchJiraIssues(
  jql: string,
  maxResults: number,
): Promise<JiraIssue[]> {
  if (isMockMode()) {
    const { SPRINT_ISSUES_BY_ID } = await import("../mockData/issues");
    const all = SPRINT_ISSUES_BY_ID[23] ?? [];
    const q = jql.toLowerCase();
    const filtered = all.filter(
      (i) =>
        i.summary.toLowerCase().includes(q) ||
        i.key.toLowerCase().includes(q) ||
        i.status.toLowerCase().includes(q),
    );
    return filtered.slice(0, maxResults);
  }
  return invoke<JiraIssue[]>("search_jira_issues", { jql, maxResults });
}

/** Diagnostic: fetch ALL fields for one issue with human-readable names.
 *  Uses ?expand=names so field IDs are mapped to display names without admin access.
 *  Returns custom fields sorted by name, standard fields first. */
export interface RawIssueField {
  id: string;
  name: string;
  value: string;
}

export async function getRawIssueFields(
  issueKey: string,
): Promise<RawIssueField[]> {
  return invoke<RawIssueField[]>("get_raw_issue_fields", { issueKey });
}

export interface JiraFieldMeta {
  id: string;
  name: string;
  fieldType: string | null;
}

/** Fetch all field definitions from the JIRA workspace (id + name + type). */
export async function getJiraFields(): Promise<JiraFieldMeta[]> {
  return invoke<JiraFieldMeta[]>("get_jira_fields");
}

export async function updateJiraIssue(
  issueKey: string,
  summary: string | null,
  description: string,
): Promise<void> {
  return invoke("update_jira_issue", { issueKey, summary, description });
}

/**
 * Update multiple fields on a JIRA issue in a single PUT request.
 * `fieldsJson` is a JSON string mapping JIRA field IDs to plain-text values.
 * e.g. { "summary": "...", "customfield_10034": "..." }
 */
export async function updateJiraFields(
  issueKey: string,
  fieldsJson: string,
): Promise<void> {
  return invoke("update_jira_fields", { issueKey, fieldsJson });
}

File: src-tauri/src/commands/jira.rs
use crate::integrations::jira::{
    CustomFieldConfig, JiraClient, JiraFieldMeta, JiraIssue, JiraSprint, RawIssueField,
};
use crate::storage::credentials::get_credential;
use crate::storage::preferences::get_pref;

fn get_config(key: &str) -> Option<String> {
    get_pref(key).or_else(|| get_credential(key))
}

fn jira_client() -> Result<(JiraClient, i64), String> {
    let base_url =
        get_credential("jira_base_url").ok_or("JIRA URL not configured. Check Settings.")?;
    let email = get_credential("jira_email").ok_or("JIRA email not configured. Check Settings.")?;
    let api_token =
        get_credential("jira_api_token").ok_or("JIRA API token not configured. Check Settings.")?;
    let board_id_str = get_config("jira_board_id")
        .ok_or("JIRA board ID not configured. Check Settings → Configuration.")?;
    let board_id: i64 = board_id_str
        .trim()
        .parse()
        .map_err(|_| "JIRA board ID must be a number. Check Settings → Configuration.")?;

    let client = JiraClient::new(base_url, email, api_token)?;
    Ok((client, board_id))
}

/// Active sprint for the configured board.
#[tauri::command]
pub async fn get_active_sprint() -> Result<Option<JiraSprint>, String> {
    let (client, board_id) = jira_client()?;
    client.get_active_sprint(board_id).await
}

/// All active sprints for the configured board (boards can have multiple active sprints).
#[tauri::command]
pub async fn get_all_active_sprints() -> Result<Vec<JiraSprint>, String> {
    let (client, board_id) = jira_client()?;
    client.get_all_active_sprints(board_id).await
}

/// All issues across all active sprints.
#[tauri::command]
pub async fn get_all_active_sprint_issues() -> Result<Vec<(JiraSprint, Vec<JiraIssue>)>, String> {
    let (client, board_id) = jira_client()?;
    let cfg = CustomFieldConfig::default();
    let sprints = client.get_all_active_sprints(board_id).await?;
    let mut result = Vec::new();
    for sprint in sprints {
        let issues = client.get_sprint_issues(sprint.id, None, &cfg).await?;
        result.push((sprint, issues));
    }
    Ok(result)
}

/// All issues in a specific sprint.
#[tauri::command]
pub async fn get_sprint_issues(sprint_id: i64) -> Result<Vec<JiraIssue>, String> {
    let (client, _) = jira_client()?;
    let cfg = CustomFieldConfig::default();
    client.get_sprint_issues(sprint_id, None, &cfg).await
}

/// All issues in the active sprint.
#[tauri::command]
pub async fn get_active_sprint_issues() -> Result<Vec<JiraIssue>, String> {
    let (client, board_id) = jira_client()?;
    let cfg = CustomFieldConfig::default();
    match client.get_active_sprint(board_id).await? {
        Some(sprint) => client.get_sprint_issues(sprint.id, None, &cfg).await,
        None => Ok(vec![]),
    }
}

/// Issues in the active sprint assigned to the authenticated user.
#[tauri::command]
pub async fn get_my_sprint_issues() -> Result<Vec<JiraIssue>, String> {
    let (client, _) = jira_client()?;
    let cfg = CustomFieldConfig::default();
    let jql = "assignee = currentUser() AND sprint in openSprints() ORDER BY priority DESC";
    client.search_issues(jql, 50, &cfg).await
}

/// Full detail for a single issue.
#[tauri::command]
pub async fn get_issue(issue_key: String) -> Result<JiraIssue, String> {
    let (client, _) = jira_client()?;
    let cfg = CustomFieldConfig::default();
    client.get_issue(&issue_key, &cfg).await
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProxiedJiraImage {
    pub content_type: String,
    pub data_base64: String,
}

/// Fetch a JIRA-hosted image (typically an attachment content URL) with the
/// configured Basic auth and return its bytes base64-encoded. Mirrors the
/// Bitbucket image proxy: the Tauri webview can't supply per-request auth
/// for `<img src>`, so this stands in. The URL is validated against the
/// configured JIRA base URL to refuse arbitrary fetches.
#[tauri::command]
pub async fn fetch_jira_image(url: String) -> Result<ProxiedJiraImage, String> {
    use base64::Engine;
    let (client, _) = jira_client()?;
    // Strict allow-list: URL must start with the configured JIRA base URL.
    // We trim any trailing slash on the base before comparison so trivial
    // formatting differences don't lock out valid attachment URLs.
    let base = client.base_url.trim_end_matches('/');
    if !url.starts_with(base) {
        return Err(format!(
            "Refusing to proxy URL outside the configured JIRA host. Expected prefix '{base}'.",
        ));
    }
    let (bytes, content_type) = client.fetch_authed_bytes(&url).await?;
    Ok(ProxiedJiraImage {
        content_type: content_type
            .unwrap_or_else(|| "application/octet-stream".to_string()),
        data_base64: base64::engine::general_purpose::STANDARD.encode(&bytes),
    })
}

/// Most-recent closed sprints, newest first.
#[tauri::command]
pub async fn get_completed_sprints(limit: u32) -> Result<Vec<JiraSprint>, String> {
    let (client, board_id) = jira_client()?;
    client.get_completed_sprints(board_id, limit as usize).await
}

/// Issues in a completed sprint by sprint ID.
#[tauri::command]
pub async fn get_sprint_issues_by_id(sprint_id: i64, complete_date: Option<String>) -> Result<Vec<JiraIssue>, String> {
    let (client, _) = jira_client()?;
    let cfg = CustomFieldConfig::default();
    client.get_sprint_issues(sprint_id, complete_date.as_deref(), &cfg).await
}

/// Future (not-yet-started) sprints for the configured board, soonest first.
#[tauri::command]
pub async fn get_future_sprints(limit: u32) -> Result<Vec<JiraSprint>, String> {
    let (client, board_id) = jira_client()?;
    client.get_future_sprints(board_id, limit as usize).await
}

/// General-purpose JQL search (used by Groom Ticket and other workflows).
#[tauri::command]
pub async fn search_jira_issues(jql: String, max_results: u32) -> Result<Vec<JiraIssue>, String> {
    let (client, _) = jira_client()?;
    let cfg = CustomFieldConfig::default();
    client.search_issues(&jql, max_results as usize, &cfg).await
}

/// Diagnostic: fetch ALL fields for a single issue (with ?expand=names so field IDs
/// are mapped to human-readable display names). No admin permissions required.
/// Returns a list of { id, name, value } objects sorted custom-fields-last.
#[tauri::command]
pub async fn get_raw_issue_fields(issue_key: String) -> Result<Vec<RawIssueField>, String> {
    let (client, _) = jira_client()?;
    client.get_raw_issue_fields(&issue_key).await
}

/// Fetch all field definitions from the JIRA workspace (id + name + type).
/// Use this alongside get_raw_issue_fields to map field IDs to human-readable names.
#[tauri::command]
pub async fn get_jira_fields() -> Result<Vec<JiraFieldMeta>, String> {
    let (client, _) = jira_client()?;
    client.get_all_fields().await
}

/// Update an issue's description (and optionally its summary).
/// `description` is plain text; the Rust layer converts it to ADF for the v3 API.
#[tauri::command]
pub async fn update_jira_issue(
    issue_key: String,
    summary: Option<String>,
    description: String,
) -> Result<(), String> {
    let (client, _) = jira_client()?;
    client
        .update_issue_description(&issue_key, summary.as_deref(), &description)
        .await
}

/// Update multiple fields on a JIRA issue in a single PUT request.
/// `fields_json` is a JSON object mapping JIRA field IDs to plain-text values.
/// Standard fields: "summary", "description".
/// Custom fields: "customfield_10034", etc. All text values are wrapped in ADF
/// doc nodes for the v3 API (except "summary" which is a plain string).
#[tauri::command]
pub async fn update_jira_fields(issue_key: String, fields_json: String) -> Result<(), String> {
    let (client, _) = jira_client()?;
    client.update_fields(&issue_key, &fields_json).await
}
